src/mjolnir/pbfgraphparser.cc

```diff
@@ -46,6 +46,94 @@ int get_number(const std::string& tag, const std::string& value) { // NOLINT
   return num;
 }
 
+// This class helps to set "culdesac" labels to loop roads correctly.
+class culdesac_processor {
+public:
+  // Adds a loop road as a candidate to be a "culdesac" road.
+  void add_candidate(uint64_t osm_way_id,
+                     size_t osm_way_index,
+                     const std::vector<uint64_t>& osm_node_ids) {
+    loops_meta_.emplace(osm_way_id, loop_meta(osm_way_index));
+    for (auto osm_node_id : osm_node_ids)
+      node_to_loop_way_.emplace(osm_node_id, osm_way_id);
+  }
+
+  // Clarifies types of loop roads and saves fixed ways.
+  void clarify_and_fix(sequence<OSMWayNode>& osm_way_node_seq, sequence<OSMWay>& osm_way_seq) {
+    osm_way_node_seq.flush();
+    osm_way_seq.flush();
+
+    size_t number_of_nodes = 0;
+    size_t count_node = 0;
+    OSMWay osm_way;
+    for (const auto& osm_way_node : osm_way_node_seq) {
+      // Reads a new way only after its nodes are read.
+      if (number_of_nodes == count_node) {
+        osm_way = *osm_way_seq[osm_way_node.way_index];
+        number_of_nodes = osm_way.node_count();
+        count_node = 0;
+      }
+
+      const auto node_to_loop_way_range = node_to_loop_way_.equal_range(osm_way_node.node.osmid_);
+      for (auto it = node_to_loop_way_range.first; it != node_to_loop_way_range.second; ++it) {
+        if (osm_way.way_id() != it->second && osm_way.use() == Use::kRoad) {
+          loops_meta_.at(it->second).add_id_of_intersection(osm_way_node.node.osmid_);
+        }
+      }
+
+      ++count_node;
+    }
+
+    fix(osm_way_seq);
+  }
+
+private:
+  // loop_meta is a helper class that stores loop info.
+  class loop_meta {
+  public:
+    explicit loop_meta(size_t way_index) : way_index_(way_index) {
+    }
+
+    size_t get_way_index() const {
+      return way_index_;
+    }
+
+    bool is_culdesac() const {
+      return node_ids_of_intersections_.size() <= 1;
+    }
+
+    void add_id_of_intersection(uint64_t node_id) {
+      node_ids_of_intersections_.insert(node_id);
+    }
+
+  private:
+    size_t way_index_;
+    // Stores nodes that are intersections of loop road loop and adjacent roads.
+    std::unordered_set<uint64_t> node_ids_of_intersections_;
+  };
+
+  // Sets "culdesac" labels to loop roads and saves ways.
+  void fix(sequence<OSMWay>& osm_way_seq) {
+    size_t number_of_culdesac = 0;
+    for (const auto& loop_way_id_to_meta : loops_meta_) {
+      const auto& meta = loop_way_id_to_meta.second;
+      if (meta.is_culdesac()) {
+        auto way_it = osm_way_seq.at(meta.get_way_index());
+        auto way = *way_it;
+        way.set_use(Use::kCuldesac);
+        way_it = way;
+        ++number_of_culdesac;
+      }
+    }
+
+    LOG_INFO("Added " + std::to_string(number_of_culdesac) + " culdesac roundabouts from " +
```

Comment:
why is it `INFO` level , not `DEBUG` ?

Reply:
I cannot imagine how it can be useful for debug
It is printed one time, I think INFO level fits for this message and frequency

Reply:
ok